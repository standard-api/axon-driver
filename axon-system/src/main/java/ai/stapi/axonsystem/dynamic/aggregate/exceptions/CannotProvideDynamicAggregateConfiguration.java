package ai.stapi.axonsystem.dynamic.aggregate.exceptions;

public class CannotProvideDynamicAggregateConfiguration extends Exception {

  public CannotProvideDynamicAggregateConfiguration(String aggregateType) {
    super(
        String.format(
            "Cannot provide Dynamic Aggregate Configuration for '%s', because it does not exists.",
            aggregateType
        )
    );
  }

  public CannotProvideDynamicAggregateConfiguration(String aggregateType, Throwable cause) {
    super(
        String.format(
            "Cannot provide Dynamic Aggregate Configuration for '%s', because it does not exists.\n Cause: %s",
            aggregateType,
            cause.getMessage()
        ),
        cause
    );
  }
}
