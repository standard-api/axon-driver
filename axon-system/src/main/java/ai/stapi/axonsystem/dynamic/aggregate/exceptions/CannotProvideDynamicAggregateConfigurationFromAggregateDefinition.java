package ai.stapi.axonsystem.dynamic.aggregate.exceptions;

import ai.stapi.graphsystem.aggregatedefinition.model.CommandHandlerDefinitionDTO;
import ai.stapi.graphsystem.operationdefinition.model.OperationDefinitionDTO;

public class CannotProvideDynamicAggregateConfigurationFromAggregateDefinition
    extends RuntimeException {

  private CannotProvideDynamicAggregateConfigurationFromAggregateDefinition(String becauseMessage) {
    super("Cannot provide DynamicAggregateConfiguration from OperationDefinition, because "
        + becauseMessage);
  }

  public static CannotProvideDynamicAggregateConfigurationFromAggregateDefinition becauseOperationHasNoneOrMoreResources(
      OperationDefinitionDTO operationDefinitionDTO
  ) {
    return new CannotProvideDynamicAggregateConfigurationFromAggregateDefinition(
        "operation definition has none or more resources." +
            "\nOperation definition: " + operationDefinitionDTO
    );
  }

  public static CannotProvideDynamicAggregateConfigurationFromAggregateDefinition becauseCommandHandlerDefinitionHasUnknownCreationalPolicy(
      CommandHandlerDefinitionDTO commandDefinition
  ) {
    return new CannotProvideDynamicAggregateConfigurationFromAggregateDefinition(
        "command handler definition has unknown creational policy." +
            "\nCommand handler definition: " + commandDefinition
    );
  }
}
