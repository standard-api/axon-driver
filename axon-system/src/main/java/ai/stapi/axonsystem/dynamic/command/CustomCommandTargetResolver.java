package ai.stapi.axonsystem.dynamic.command;

import ai.stapi.graphsystem.messaging.command.AbstractCommand;
import org.axonframework.commandhandling.CommandMessage;
import org.axonframework.modelling.command.AnnotationCommandTargetResolver;
import org.axonframework.modelling.command.CommandTargetResolver;
import org.axonframework.modelling.command.VersionedAggregateIdentifier;
import org.jetbrains.annotations.NotNull;

public class CustomCommandTargetResolver implements CommandTargetResolver {
  
  private final AnnotationCommandTargetResolver annotationCommandTargetResolver;

  public CustomCommandTargetResolver() {
    this.annotationCommandTargetResolver = AnnotationCommandTargetResolver.builder().build();
  }

  @Override
  public VersionedAggregateIdentifier resolveTarget(@NotNull CommandMessage<?> command) {
    var payload = command.getPayload();
    if (payload instanceof AbstractCommand<?> abstractCommand) {
      return new VersionedAggregateIdentifier(
          abstractCommand.getTargetIdentifier(),
          null
      );
    }
    return this.annotationCommandTargetResolver.resolveTarget(command);
  }
}
