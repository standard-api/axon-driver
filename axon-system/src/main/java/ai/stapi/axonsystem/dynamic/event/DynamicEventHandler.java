package ai.stapi.axonsystem.dynamic.event;

import ai.stapi.axonsystem.dynamic.messagehandler.DynamicMessageHandler;
import ai.stapi.graphsystem.messaging.event.DynamicGraphUpdatedEvent;
import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import org.axonframework.eventhandling.EventMessage;

@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.ANNOTATION_TYPE, ElementType.CONSTRUCTOR, ElementType.METHOD})
@DynamicMessageHandler(
    messageType = EventMessage.class,
    payloadType = DynamicGraphUpdatedEvent.class
)
public @interface DynamicEventHandler {

  /**
   * Specifies the name of message that can be handled by the member method.
   */
  String messageName();

}
