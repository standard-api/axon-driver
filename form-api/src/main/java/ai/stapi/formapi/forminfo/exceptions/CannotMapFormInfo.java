package ai.stapi.formapi.forminfo.exceptions;

public class CannotMapFormInfo extends RuntimeException {

    private static final String MSG = "Cannot map form info, because ";

    private CannotMapFormInfo(String becauseMessage) {
        super(MSG + becauseMessage);
    }

    private CannotMapFormInfo(String becauseMessage, Throwable cause) {
        super(MSG + becauseMessage, cause);
    }

    public static CannotMapFormInfo becauseThereWasNoCommandHandlerDefinitionForProvidedOperation(
        String operationId
    ) {
        return new CannotMapFormInfo(
            String.format(
                "there was not command handler definition with provided operation.%nOperation name: '%s'",
                operationId
            )
        );
    }

    public static CannotMapFormInfo becauseThereWasNoCommandHandlerDefinitionForProvidedOperation(
        String operationId,
        Throwable cause
    ) {
        return new CannotMapFormInfo(
            String.format(
                "there was not command handler definition with provided operation.%nOperation name: '%s'",
                operationId
            ),
            cause
        );
    }
}
