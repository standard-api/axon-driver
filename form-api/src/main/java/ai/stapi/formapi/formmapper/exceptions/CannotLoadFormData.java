package ai.stapi.formapi.formmapper.exceptions;

public class CannotLoadFormData extends RuntimeException {

    private static final String MSG = "Cannot load data for form, because ";

    private CannotLoadFormData(String becauseMessage) {
        super(MSG + becauseMessage);
    }

    private CannotLoadFormData(String becauseMessage, Throwable cause) {
        super(MSG + becauseMessage, cause);
    }

    public static CannotLoadFormData becauseThereWasNoCommandHandlerDefinitionForProvidedOperation(
        String operationId
    ) {
        return new CannotLoadFormData(
            String.format(
                "there was not command handler definition with provided operation.%nOperation name: '%s'",
                operationId
            )
        );
    }

    public static CannotLoadFormData becauseThereWasNoCommandHandlerDefinitionForProvidedOperation(
        String operationId,
        Throwable cause
    ) {
        return new CannotLoadFormData(
            String.format(
                "there was not command handler definition with provided operation.%nOperation name: '%s'",
                operationId
            ),
            cause
        );
    }

    public static CannotLoadFormData becauseAggregateByProvidedResourceIdWasNotFound(
        String resourceId,
        Throwable cause
    ) {
        return new CannotLoadFormData(
            String.format(
                "no aggregate with provided resource id was found.%nResource id: '%s'",
                resourceId
            ),
            cause
        );
    }
}
