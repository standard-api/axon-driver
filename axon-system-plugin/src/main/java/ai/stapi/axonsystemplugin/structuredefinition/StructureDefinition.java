package ai.stapi.axonsystemplugin.structuredefinition;

import ai.stapi.axonsystem.graphaggregate.AggregateWithDynamicGraph;
import ai.stapi.graphoperations.objectGraphMapper.model.MissingFieldResolvingStrategy;
import ai.stapi.graphsystem.dynamiccommandprocessor.DynamicCommandProcessor;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.FixedImportStructureDefinition;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.ImportStructureDefinition;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.StructureDefinitionImported;
import ai.stapi.schema.structuredefinition.StructureDefinitionId;
import ai.stapi.schema.structuredefinition.StructureDefinitionNormalizer;
import org.axonframework.commandhandling.CommandHandler;
import org.axonframework.eventsourcing.EventSourcingHandler;
import org.axonframework.modelling.command.AggregateCreationPolicy;
import org.axonframework.modelling.command.CreationPolicy;
import org.axonframework.spring.stereotype.Aggregate;
import org.springframework.beans.factory.annotation.Autowired;

@Aggregate
public class StructureDefinition extends AggregateWithDynamicGraph<StructureDefinitionId> {

  @CommandHandler
  @CreationPolicy(AggregateCreationPolicy.CREATE_IF_MISSING)
  public void handle(
      ImportStructureDefinition command,
      @Autowired DynamicCommandProcessor commandProcessor
  ) {
    this.commandProcessor = commandProcessor;
    this.processCommandDynamically(
        new FixedImportStructureDefinition(
            command.getTargetIdentifier(),
            StructureDefinitionNormalizer.normalize(command.getStructureDefinitionSource())
        ),
        MissingFieldResolvingStrategy.LENIENT
    );
  }

  @EventSourcingHandler
  public void on(StructureDefinitionImported event) {
    this.onAggregateCreated(event);
  }
}