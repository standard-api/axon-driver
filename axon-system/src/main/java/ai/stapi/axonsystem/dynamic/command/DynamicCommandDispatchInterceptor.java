package ai.stapi.axonsystem.dynamic.command;

import ai.stapi.graphsystem.messaging.command.DynamicCommand;
import java.util.List;
import java.util.function.BiFunction;
import org.axonframework.commandhandling.CommandMessage;
import org.axonframework.commandhandling.GenericCommandMessage;
import org.axonframework.messaging.GenericMessage;
import org.axonframework.messaging.MessageDispatchInterceptor;
import org.jetbrains.annotations.NotNull;

public class DynamicCommandDispatchInterceptor  implements MessageDispatchInterceptor<CommandMessage<?>> {

  @NotNull
  @Override
  public BiFunction<Integer, CommandMessage<?>, CommandMessage<?>> handle(
      @NotNull List<? extends CommandMessage<?>> messages
  ) {
    return (integer, commandMessage) -> {
      if (commandMessage.getPayload() instanceof DynamicCommand dynamicCommand) {
        return new GenericCommandMessage<>(
            new GenericMessage<>(dynamicCommand),
            dynamicCommand.getSerializationType()
        );
      }
      return commandMessage;
    };
  }
}
