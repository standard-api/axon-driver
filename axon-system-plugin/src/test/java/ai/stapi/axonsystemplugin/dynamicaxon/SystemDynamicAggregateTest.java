package ai.stapi.axonsystemplugin.dynamicaxon;

import ai.stapi.graphsystem.aggregategraphstatemodifier.exceptions.CannotAddToAggregateState;
import ai.stapi.graphsystem.messaging.command.DynamicCommand;
import ai.stapi.graphsystem.messaging.event.DynamicGraphUpdatedEvent;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.ImportStructureDefinition;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.StructureDefinitionImported;
import ai.stapi.schema.structuredefinition.RawStructureDefinitionData;
import ai.stapi.schema.structuredefinition.RawStructureDefinitionElementDefinition;
import ai.stapi.schema.structuredefinition.RawStructureDefinitionElementDefinition.ElementDefinitionType;
import ai.stapi.schema.structuredefinition.StructureDefinitionId;
import ai.stapi.test.domain.DomainTestCase;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.apache.commons.io.FileUtils;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;
import org.springframework.beans.factory.annotation.Autowired;

class SystemDynamicAggregateTest extends DomainTestCase {
  
  @Autowired
  private ObjectMapper objectMapper;

  @Test
  void itCannotAddItemToListFieldIfTheSubnodeTobeModifiedDoesntExist() throws IOException {
    this.givenStructureForElementWithoutDifferentialCreated();
    var command = new DynamicCommand(
        new StructureDefinitionId("Element"),
        "AddElementOnStructureDefinitionDifferential",
        new HashMap<>(Map.of(
            "element", new ArrayList<>(List.of(
                new RawStructureDefinitionElementDefinition(
                    "Element.newField",
                    0,
                    "1",
                    "A newly added element",
                    "A newly added element",
                    "A newly added element",
                    new ArrayList<>(List.of(
                        new ElementDefinitionType(
                            "string",
                            new ArrayList<>()
                        )
                    ))
                )
            ))
        ))
    );
    Executable executable = () -> this.whenCommandIsDispatched(command);
    this.thenExceptionMessageApproved(CannotAddToAggregateState.class, executable);
  }

  @Test
  void itCanAddItemToListFieldOnSubnode() throws IOException {
    this.givenStructureForElementCreated();
    var command = new DynamicCommand(
        new StructureDefinitionId("Element"),
        "AddElementOnStructureDefinitionDifferential",
        new HashMap<>(Map.of(
            "element", new ArrayList<>(List.of(
                new RawStructureDefinitionElementDefinition(
                    "Element.newField",
                    0,
                    "1",
                    "A newly added element",
                    "A newly added element",
                    "A newly added element",
                    new ArrayList<>(List.of(
                        new ElementDefinitionType(
                            "string",
                            new ArrayList<>()
                        )
                    ))
                )
            ))
        ))
    );
    this.whenCommandIsDispatched(command);
    this.thenExpectedEventTypesSaved(
        StructureDefinitionImported.class,
        DynamicGraphUpdatedEvent.class
    );
    this.thenLastDynamicEventOfNameApproved("ElementOnStructureDefinitionDifferentialAdded");
  }

  @Test
  void itCanAddMultipleItemsToListFieldOnSubnode() throws IOException {
    this.givenStructureForElementCreated();
    var command = new DynamicCommand(
        new StructureDefinitionId("Element"),
        "AddElementOnStructureDefinitionDifferential",
        new HashMap<>(Map.of(
            "element", new ArrayList<>(List.of(
                new RawStructureDefinitionElementDefinition(
                    "Element.newField",
                    0,
                    "1",
                    "A newly added element",
                    "A newly added element",
                    "A newly added element",
                    new ArrayList<>(List.of(
                        new ElementDefinitionType(
                            "string",
                            new ArrayList<>()
                        )
                    ))
                ),
                new RawStructureDefinitionElementDefinition(
                    "Element.anotherField",
                    0,
                    "1",
                    "A anotherField added element",
                    "A anotherField added element",
                    "A anotherField added element",
                    new ArrayList<>(List.of(
                        new ElementDefinitionType(
                            "string",
                            new ArrayList<>()
                        )
                    ))
                )
            ))
        ))
    );
    this.whenCommandIsDispatched(command);
    this.thenExpectedEventTypesSaved(
        StructureDefinitionImported.class,
        DynamicGraphUpdatedEvent.class
    );
    this.thenLastDynamicEventOfNameApproved("ElementOnStructureDefinitionDifferentialAdded");
  }

  @Test
  void itCannotAddItemToListFieldLocatedInListOfSubnodesWhenIdOfSubnodeNotFound()
      throws IOException {
    this.givenStructureForElementCreated();
    var command = new DynamicCommand(
        new StructureDefinitionId("Element"),
        "AddCodingOnStructureDefinitionJurisdiction",
        new HashMap<>(Map.of(
            "coding", new ArrayList<>(List.of(
                new HashMap<>(Map.of(
                    "code", "Some Jurisdiction Code"
                ))
            )),
            "jurisdictionId", "Jurisdiction/NotExistingJurisdiction"
        ))
    );
    Executable executable = () -> this.whenCommandIsDispatched(command);
    this.thenExceptionMessageApproved(CannotAddToAggregateState.class, executable);
  }

  @Test
  void itCanAddItemToListFieldLocatedInListOfSubnodes() throws IOException {
    this.givenStructureForElementCreated();
    var jurisdictionId = "JurisdictionId";
    var structureId = new StructureDefinitionId("Element");
    var givenCommand = new DynamicCommand(
        structureId,
        "AddJurisdictionOnStructureDefinition",
        new HashMap<>(Map.of(
            "jurisdiction", new ArrayList<>(List.of(
                new HashMap<>(Map.of(
                    "id", jurisdictionId,
                    "text", "Some Jurisdiction Text"
                ))
            ))
        ))
    );
    this.givenCommandIsDispatched(givenCommand);
    var command = new DynamicCommand(
        structureId,
        "AddCodingOnStructureDefinitionJurisdiction",
        new HashMap<>(Map.of(
            "coding", new ArrayList<>(List.of(
                new HashMap<>(Map.of(
                    "code", "Some Jurisdiction Code"
                ))
            )),
            "jurisdictionId", String.format("CodeableConcept/%s", jurisdictionId)
        ))
    );
    this.whenCommandIsDispatched(command);
    this.thenExpectedEventTypesSaved(
        StructureDefinitionImported.class,
        DynamicGraphUpdatedEvent.class,
        DynamicGraphUpdatedEvent.class
    );
    this.thenExpectedDynamicEventsSaved(
        "JurisdictionOnStructureDefinitionAdded",
        "CodingOnStructureDefinitionJurisdictionAdded"
    );
    this.thenMergedGraphOfAggregateApproved(structureId);
  }

  @Test
  void itCanAddItemToListFieldLocatedInListFieldOfListOfSubnodes() throws IOException {
    //Given
    this.givenStructureForElementCreated();
    var structureId = new StructureDefinitionId("Element");
    var exampleElementId = "ExampleElementId";
    var addElementCommand = new DynamicCommand(
        structureId,
        "AddElementOnStructureDefinitionDifferential",
        new HashMap<>(Map.of(
            "element", new ArrayList<>(List.of(
                Map.of(
                    "id", exampleElementId,
                    "path", "Element.newField",
                    "min", 0,
                    "max", "1",
                    "shortDescription", "A newly added element",
                    "definition", "A newly added element",
                    "comment", "A newly added element",
                    "type", new ArrayList<>(List.of(
                        new ElementDefinitionType(
                            "string",
                            new ArrayList<>()
                        )
                    ))
                )
            ))
        ))
    );
    this.givenCommandIsDispatched(addElementCommand);
    var exampleTypeId = "ExampleTypeId";
    var addElementTypeCommand = new DynamicCommand(
        structureId,
        "AddTypeOnStructureDefinitionDifferentialElement",
        new HashMap<>(Map.of(
            "elementId", String.format("ElementDefinition/%s", exampleElementId),
            "type", new ArrayList<>(List.of(
                Map.of(
                    "id", exampleTypeId,
                    "code", "Reference"
                )
            ))
        ))
    );
    this.givenCommandIsDispatched(addElementTypeCommand);
    var command = new DynamicCommand(
        structureId,
        "AddTargetProfileOnStructureDefinitionDifferentialElementType",
        new HashMap<>(Map.of(
            "typeId", String.format("ElementDefinitionType/%s", exampleTypeId),
            "targetProfile", new ArrayList<>(List.of(
                "SomeTargetProfile"
            ))
        ))
    );
    this.whenCommandIsDispatched(command);
    this.thenExpectedEventTypesSaved(
        StructureDefinitionImported.class,
        DynamicGraphUpdatedEvent.class,
        DynamicGraphUpdatedEvent.class,
        DynamicGraphUpdatedEvent.class
    );
    this.thenExpectedDynamicEventsSaved(
        "ElementOnStructureDefinitionDifferentialAdded",
        "TypeOnStructureDefinitionDifferentialElementAdded",
        "TargetProfileOnStructureDefinitionDifferentialElementTypeAdded"
    );
    this.thenMergedGraphOfAggregateApproved(structureId);
  }

  //TODO: uncomment after aggregate definition loaders
//  @Test
//  void itCanAddItemToListFieldOfUnionMember() {
//    var structureId = new UniqueIdentifier("ExampleId");
//    this.givenCommandIsDispatched(
//        new DynamicCommand(
//            structureId,
//            "CreateExampleDynamicAggregateTypeWithUnion",
//            Map.of(
//                "exampleUnionField", new HashMap<>(Map.of(
//                    "serializationType", "Timing",
//                    "code", new HashMap<>(Map.of(
//                        "text", "Some Irrelevant Timing code text"
//                    ))
//                ))
//            )
//        )
//    );
//    this.whenCommandIsDispatched(
//        new DynamicCommand(
//            structureId,
//            "AddEventOnExampleDynamicAggregateTypeWithUnionExampleUnionField",
//            Map.of(
//                "exampleUnionField", new HashMap<>(Map.of(
//                    "event", "Pica"
//                ))
//            )
//        )
//    );
//    this.thenExpectedDynamicEventsSaved(
//        "ElementOnStructureDefinitionDifferentialAdded",
//        "TypeOnStructureDefinitionDifferentialElementAdded",
//        "TargetProfileOnStructureDefinitionDifferentialElementTypeAdded"
//    );
//    this.thenMergedGraphOfAggregateApproved(structureId);
//  }

  private void givenStructureForElementCreated() throws IOException {
    var fixtureFilePath = this.getFixtureFilePath("element.profile.canonical.json");
    var file = new File(fixtureFilePath);
    var fileContent = FileUtils.readFileToString(file, "UTF-8");
    RawStructureDefinitionData structureDefinitionSource;
    try {
      structureDefinitionSource = this.objectMapper.readValue(fileContent, RawStructureDefinitionData.class);
    } catch (JsonProcessingException e) {
      throw new RuntimeException(e);
    }
    var givenCommand = new ImportStructureDefinition(
        new StructureDefinitionId("Element"),
        structureDefinitionSource
    );
    this.givenCommandIsDispatched(givenCommand);
  }

  private void givenStructureForElementWithoutDifferentialCreated() throws IOException {
    var fixtureFilePath = this.getFixtureFilePath("element.profile.canonical.json");
    var file = new File(fixtureFilePath);
    var fileContent = FileUtils.readFileToString(file, "UTF-8");
    RawStructureDefinitionData structureDefinitionSource;
    try {
      structureDefinitionSource =
          this.objectMapper.readValue(fileContent, RawStructureDefinitionData.class);
    } catch (JsonProcessingException e) {
      throw new RuntimeException(e);
    }
    structureDefinitionSource = new RawStructureDefinitionData(
        structureDefinitionSource.getId(),
        structureDefinitionSource.getUrl(),
        structureDefinitionSource.getStatus(),
        structureDefinitionSource.getDescription(),
        structureDefinitionSource.getKind(),
        structureDefinitionSource.getIsAbstract(),
        structureDefinitionSource.getType(),
        structureDefinitionSource.getBaseDefinition(),
        null
    );
    var givenCommand = new ImportStructureDefinition(
        new StructureDefinitionId("Element"),
        structureDefinitionSource
    );
    this.givenCommandIsDispatched(givenCommand);
  }
}
