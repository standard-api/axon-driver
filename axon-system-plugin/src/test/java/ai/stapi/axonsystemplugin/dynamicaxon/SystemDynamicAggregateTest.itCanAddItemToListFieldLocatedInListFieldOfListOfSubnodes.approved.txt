nodes:
	node_type: ElementDefinition
	node_hash: 1388F
	node_edges:
		ElementDefinition -> type -> ElementDefinitionType (1388F -> 29DBB -> FB86E)
		StructureDefinitionDifferential -> element -> ElementDefinition (47268 -> 4E804 -> 1388F)
	attributes:
		definition -> Unique id for the element within a resource (for internal references). This may be any string value that does not contain spaces.
		max -> 1
		min -> 0
		path -> Element.id
		short -> Unique id for inter-element referencing

	node_type: ElementDefinition
	node_hash: 52A6D
	node_edges:
		ElementDefinition -> type -> ElementDefinitionType (52A6D -> 2AE97 -> FEC55)
		ElementDefinition -> type -> ElementDefinitionType (52A6D -> BCB0A -> 2687C)
		StructureDefinitionDifferential -> element -> ElementDefinition (47268 -> 64991 -> 52A6D)
	attributes:
		comment -> A newly added element
		definition -> A newly added element
		max -> 1
		min -> 0
		path -> Element.newField

	node_type: ElementDefinition
	node_hash: AE262
	node_edges:
		ElementDefinition -> type -> ElementDefinitionType (AE262 -> 771A8 -> B3F79)
		StructureDefinitionDifferential -> element -> ElementDefinition (47268 -> 77622 -> AE262)
	attributes:
		comment -> There can be no stigma associated with the use of extensions by any application, project, or standard - regardless of the institution or jurisdiction that uses or defines the extensions.  The use of extensions is what allows the FHIR specification to retain a core level of simplicity for everyone.
		definition -> May be used to represent additional information that is not part of the basic definition of the element. To make the use of extensions safe and manageable, there is a strict set of governance  applied to the definition and use of extensions. Though any implementer can define an extension, there is a set of requirements that SHALL be met as part of the definition of the extension.
		max -> *
		min -> 0
		path -> Element.extension
		short -> Additional content defined by implementations

	node_type: ElementDefinition
	node_hash: DD3CE
	node_edges:
		StructureDefinitionDifferential -> element -> ElementDefinition (47268 -> 090EC -> DD3CE)
	attributes:
		definition -> Base definition for all elements in a resource.
		max -> *
		min -> 0
		path -> Element
		short -> Base for all elements

	node_type: ElementDefinitionType
	node_hash: 2687C
	node_edges:
		ElementDefinition -> type -> ElementDefinitionType (52A6D -> BCB0A -> 2687C)
	attributes:
		code -> Reference
		targetProfile -> [SomeTargetProfile]

	node_type: ElementDefinitionType
	node_hash: B3F79
	node_edges:
		ElementDefinition -> type -> ElementDefinitionType (AE262 -> 771A8 -> B3F79)
	attributes:
		code -> Extension
		targetProfile -> []

	node_type: ElementDefinitionType
	node_hash: FB86E
	node_edges:
		ElementDefinition -> type -> ElementDefinitionType (1388F -> 29DBB -> FB86E)
	attributes:
		code -> string
		targetProfile -> []

	node_type: ElementDefinitionType
	node_hash: FEC55
	node_edges:
		ElementDefinition -> type -> ElementDefinitionType (52A6D -> 2AE97 -> FEC55)
	attributes:
		code -> string
		targetProfile -> []

	node_type: StructureDefinition
	node_hash: 955AE
	node_edges:
		StructureDefinition -> differential -> StructureDefinitionDifferential (955AE -> 378D2 -> 47268)
	attributes:
		abstract -> true
		description -> Base StructureDefinition for Element Type: Base definition for all elements in a resource.
		kind -> complex-type
		status -> active
		type -> Element
		url -> http://hl7.org/fhir/StructureDefinition/Element

	node_type: StructureDefinitionDifferential
	node_hash: 47268
	node_edges:
		StructureDefinition -> differential -> StructureDefinitionDifferential (955AE -> 378D2 -> 47268)
		StructureDefinitionDifferential -> element -> ElementDefinition (47268 -> 090EC -> DD3CE)
		StructureDefinitionDifferential -> element -> ElementDefinition (47268 -> 4E804 -> 1388F)
		StructureDefinitionDifferential -> element -> ElementDefinition (47268 -> 64991 -> 52A6D)
		StructureDefinitionDifferential -> element -> ElementDefinition (47268 -> 77622 -> AE262)
	attributes:
		parent -> Element

edges:
	edge_type: differential
	edge_hash: 378D2
	StructureDefinition -> differential -> StructureDefinitionDifferential (955AE -> 378D2 -> 47268)
	attributes:

	edge_type: element
	edge_hash: 090EC
	StructureDefinitionDifferential -> element -> ElementDefinition (47268 -> 090EC -> DD3CE)
	attributes:

	edge_type: element
	edge_hash: 4E804
	StructureDefinitionDifferential -> element -> ElementDefinition (47268 -> 4E804 -> 1388F)
	attributes:

	edge_type: element
	edge_hash: 64991
	StructureDefinitionDifferential -> element -> ElementDefinition (47268 -> 64991 -> 52A6D)
	attributes:

	edge_type: element
	edge_hash: 77622
	StructureDefinitionDifferential -> element -> ElementDefinition (47268 -> 77622 -> AE262)
	attributes:

	edge_type: type
	edge_hash: 29DBB
	ElementDefinition -> type -> ElementDefinitionType (1388F -> 29DBB -> FB86E)
	attributes:

	edge_type: type
	edge_hash: 2AE97
	ElementDefinition -> type -> ElementDefinitionType (52A6D -> 2AE97 -> FEC55)
	attributes:

	edge_type: type
	edge_hash: 771A8
	ElementDefinition -> type -> ElementDefinitionType (AE262 -> 771A8 -> B3F79)
	attributes:

	edge_type: type
	edge_hash: BCB0A
	ElementDefinition -> type -> ElementDefinitionType (52A6D -> BCB0A -> 2687C)
	attributes:
